
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
        
      
      
      <link rel="icon" href="images/icon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>TWILIGHT</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="deep-orange" data-md-color-accent="deep-purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#welcome-to-twilight-wiki" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="TWILIGHT" class="md-header__button md-logo" aria-label="TWILIGHT" data-md-component="logo">
      
  <img src="images/icon.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TWILIGHT
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Home
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="deep-orange" data-md-color-accent="deep-purple"  aria-label="Light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="deep-orange"  aria-label="Dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/TurakhiaLab/TWILIGHT" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    TurakhiaLab/TWILIGHT
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="index.html" class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="TWILIGHT" class="md-nav__button md-logo" aria-label="TWILIGHT" data-md-component="logo">
      
  <img src="images/icon.png" alt="logo">

    </a>
    TWILIGHT
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/TurakhiaLab/TWILIGHT" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    TurakhiaLab/TWILIGHT
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="index.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#twilight-video-tutorial" class="md-nav__link">
    <span class="md-ellipsis">
      
        TWILIGHT Video Tutorial
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      
        Introduction
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Introduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Overview
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#key-features" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Features
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Key Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#divide-and-conquer-technique" class="md-nav__link">
    <span class="md-ellipsis">
      
        Divide-and-Conquer Technique
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#progressive-alignment-and-scheduling" class="md-nav__link">
    <span class="md-ellipsis">
      
        Progressive Alignment and Scheduling
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#removing-gappy-columns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Removing gappy columns
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#banded-alignment-and-tiling-strategy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Banded Alignment and Tiling Strategy
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parallelization-techniques" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parallelization Techniques
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation-methods" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation Methods
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Installation Methods">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#installation-summary-choose-your-installation-method" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation summary (choose your installation method)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-conda" class="md-nav__link">
    <span class="md-ellipsis">
      
        Using Conda
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-installation-script-requires-sudo-access" class="md-nav__link">
    <span class="md-ellipsis">
      
        Using installation script (requires sudo access)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-dockerfile" class="md-nav__link">
    <span class="md-ellipsis">
      
        Using Dockerfile
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-twilight" class="md-nav__link">
    <span class="md-ellipsis">
      
        Run TWILIGHT
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Run TWILIGHT">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#twilight-command-line-interface-cli" class="md-nav__link">
    <span class="md-ellipsis">
      
        TWILIGHT Command Line Interface (CLI)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="TWILIGHT Command Line Interface (CLI)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#main-functionalities" class="md-nav__link">
    <span class="md-ellipsis">
      
        Main Functionalities
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      
        Command Line Options
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#usages-and-examples" class="md-nav__link">
    <span class="md-ellipsis">
      
        Usages and Examples
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Usages and Examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#build-msa-from-unaligned-sequences-default-mode" class="md-nav__link">
    <span class="md-ellipsis">
      
        Build MSA from unaligned Sequences (Default Mode)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merge-multiple-msas" class="md-nav__link">
    <span class="md-ellipsis">
      
        Merge Multiple MSAs
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-new-sequences-to-an-existing-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      
        Add New Sequences to an Existing Alignment
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#divide-and-conquer-method" class="md-nav__link">
    <span class="md-ellipsis">
      
        Divide-and-Conquer Method
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#flexible-tree-support" class="md-nav__link">
    <span class="md-ellipsis">
      
        Flexible Tree Support
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#snakemake-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Snakemake Workflow
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Snakemake Workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#main-functionalities_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Main Functionalities
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#command-line-configurations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Command Line Configurations
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#usages-and-examples_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Usages and Examples
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Usages and Examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#iterative-mode" class="md-nav__link">
    <span class="md-ellipsis">
      
        Iterative Mode
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-new-sequences-to-an-existing-alignment-placement-mode" class="md-nav__link">
    <span class="md-ellipsis">
      
        Add New Sequences to an Existing Alignment (Placement Mode)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contributions" class="md-nav__link">
    <span class="md-ellipsis">
      
        Contributions
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citing-twilight" class="md-nav__link">
    <span class="md-ellipsis">
      
        Citing TWILIGHT
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              
                



  


              
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/TurakhiaLab/TWILIGHT/edit/master/docs/index.md" title="Edit this page" class="md-content__button md-icon" rel="edit">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/TurakhiaLab/TWILIGHT/raw/master/docs/index.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.2 8.2 0 0 1-1.23-2"/></svg>
    </a>
  


<h1 id="welcome-to-twilight-wiki"><b>Welcome to TWILIGHT Wiki</b></h1>
<div align="center">
    <img src="images/logo.png"/>
</div>

<h2 id="twilight-video-tutorial"><b>TWILIGHT Video Tutorial</b></h2>
<iframe width="1000" height="600" src="https://www.youtube.com/embed/bn8N_Y-_g_A?si=unLXo8xNFAiNDa-K" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>

<h2 id="introduction"><b>Introduction</b></h2>
<h3 id="overview"><b>Overview</b><a name="overview"></a></h3>
<p>TWILIGHT (<strong>T</strong>all and <strong>Wi</strong>de A<strong>lig</strong>nments at <strong>H</strong>igh <strong>T</strong>hroughput) is an innovative MSA tool designed to leverage massive parallelism for efficient handling of tall and wide alignment tasks. It is able to scale to millions of long nucleotide sequences (&gt;10000 bases). TWILIGHT can run on CPU-only platforms (Linux/Mac) or take advantage of CUDA-capable GPUs for further acceleration.</p>
<p>By default, TWILIGHT requires an unaligned sequence file in FASTA format and an input guide tree in Newick format to generate the output alignment in FASTA format (Fig. 1a, see <a href="#default"><strong>TWILIGHT Default Mode</strong></a> for more details). When a guide tree is unavailable, users can utilize the iterative mode, which provides a Snakemake workflow to estimate guide trees using external tools (Fig. 1b, see <a href="#iterative"><strong>TWILIGHT Iterative Mode</strong></a> for more details).</p>
<p>TWILIGHT adopts the <strong>progressive alignment algorithm (Fig. 1c)</strong> and employs <strong>tiling strategies to band alignments (Fig. 1e)</strong>. Combined with a <strong>divide-and-conquer technique (Fig. 1a)</strong>, <strong>a novel heuristic dealing with gappy columns (Fig. 1d)</strong> and <strong>support for GPU acceleration (Fig. 1f)</strong>, TWILIGHT demonstrates exceptional speed and memory efficiency.  </p>
<div align="center">
    <div><b>Figure 1: Overview of TWILIGHT algorithm</b></div>
    <img src="images/overview.png" width="1000"/>
</div>

<h3 id="key-features"><b>Key Features</b></h3>
<h4 id="divide-and-conquer-technique"><b>Divide-and-Conquer Technique</b></h4>
<p>TWILIGHT's <strong><em>divide-and-conquer technique (Fig. 1a)</em></strong> starts by dividing the initial guide tree into smaller subtrees with at most <em>m</em> leaves, which controls memory usage by limiting the number of sequences processed at once. TWILIGHT sequentially iterates over each subtree, loading its corresponding sequences into memory, and computing the <em>subalignment</em> for the subtree. After all <em>subalignments</em> are computed, TWILIGHT merges <em>subalignments</em> with transitivity merger algorithm of PASTA (Mirarab et al., 2015) or progressive alignment.  </p>
<h4 id="progressive-alignment-and-scheduling"><b>Progressive Alignment and Scheduling</b></h4>
<p>TWILIGHT performs progressive alignment based on the guide tree structure, first
aligning the most similar sequences or sequence groups and then progressively adding less similar sequences or groups. The <strong><em>scheduling algorithm (Fig. 1c)</em></strong> initializes the leaf nodes and then performs a post-order traversal to set the alignment order for internal nodes. This approach allows certain child nodes (e.g., a, b, d, f) to be aligned in parallel, while others (e.g., c, e, g) are processed sequentially, ensuring efficient parallel processing while respecting tree dependencies.  </p>
<h4 id="removing-gappy-columns"><b>Removing gappy columns</b></h4>
<p>As more sequences are added, the alignment tends to grow longer, often with many gap-dominated columns, leading to increased computational time. To improve efficiency, TWILIGHT <strong><em>excludes gappy columns (Fig. 1d)</em></strong>‚Äîwhere gaps exceed a specified threshold‚Äîbefore the alignment step, resulting in shorter profiles and reducing computational overhead. After the alignment step, TWILIGHT restores the excluded gappy columns. If gappy columns from both profiles align at the same position, they are merged, reducing the overall alignment length. Otherwise, a gappy column from one profile is aligned with a new gap introduced in the other
profile.  </p>
<h4 id="banded-alignment-and-tiling-strategy"><b>Banded Alignment and Tiling Strategy</b></h4>
<p>TWILIGHT employs a modified X-Drop algorithm to control memory usage in pairwise profile alignment, ensuring linear scaling with sequence length. However, storing traceback pointers in limited GPU shared memory still poses challenges. To address this, TWILIGHT integrates the <strong><em>TALCO tiling algorithm  (Walia et al., 2024) (Fig. 1e)</em></strong>, which limits traceback storage to a constant memory usage by using convergence pointers.  </p>
<h4 id="parallelization-techniques"><b>Parallelization Techniques</b></h4>
<p>TWILIGHT employs parallelization techniques on both CPU and GPU to maximize efficiency. On CPUs, it is implemented in C++ with Threading Building Blocks (TBB) to exploit thread-level parallelism, processing independent alignments in parallel while also parallelizing profile calculations and sequence updates. On GPUs, TWILIGHT utilizes <strong><em>three levels of parallelism (Fig. 1f)</em></strong>: multi-GPU parallelism distributes alignment tasks into batches and sends to multiple GPUs, inter-alignment parallelism assigns each GPU thread block to a pairwise profile alignment, and intra-alignment parallelism processes multiple cells in the dynamic programming matrix wavefront simultaneously.  </p>
<p><a name="install"></a></p>
<h2 id="installation-methods"><b>Installation Methods</b></h2>
<h3 id="installation-summary-choose-your-installation-method"><strong>Installation summary (choose your installation method)</strong></h3>
<p>TWILIGHT offers multiple installation methods for different platforms and hardware setups:</p>
<ul>
<li>Conda is recommended for most users needing the <a href="#overview">default mode</a> and <em>partial</em> <a href="#overview">iterative mode</a> support, as some tree tools may be unavailable on certain platforms.</li>
<li>Install script is required for AMD GPU support.</li>
<li>Docker (built from the provided Dockerfile) is recommended for full support for <a href="#overview">iterative mode</a>.</li>
</ul>
<table>
<thead>
<tr>
<th>Platform / Setup</th>
<th><a href="#conda">Conda</a></th>
<th><a href="#script">Script</a></th>
<th><a href="#docker">Docker</a></th>
</tr>
</thead>
<tbody>
<tr>
<td>Linux (x86_64)</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
</tr>
<tr>
<td>Linux (aarch64)</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
<td>üü°</td>
</tr>
<tr>
<td>macOS (Intel Chip)</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
</tr>
<tr>
<td>macOS (Apple Silicon)</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
<td>üü°</td>
</tr>
<tr>
<td>NVIDIA GPU</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
<td>‚úÖ</td>
</tr>
<tr>
<td>AMD GPU</td>
<td>‚ùå</td>
<td>‚úÖ</td>
<td>‚ùå</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>üü° The Docker image is currently built for the <code>linux/amd64</code> platform. While it can run on <code>arm64</code> systems (e.g., Apple Silicon or Linux aarch64) via emulation, this may lead to reduced performance.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>‚ö†Ô∏è To enable <strong>GPU support</strong>, the appropriate GPU drivers must be installed on the host system. This applies to all installation methods (Installation script, Conda, and Docker). The CUDA toolkits and libraries are included in Conda and Docker setups, but must be installed manually when using the installation script.  </p>
</div>
<h3 id="using-conda"><strong>Using Conda</strong> <a name="conda"></a></h3>
<p>TWILIGHT is available on multiple platforms via Conda. See <a href="https://anaconda.org/bioconda/twilight">TWILIGHT Bioconda Page</a> for details.  </p>
<ol>
<li><a name="installconda"></a> Install Conda (if not installed)
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="c1"># Replace &lt;PLATFORM&gt; with the actual platform of your system</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="c1"># See https://repo.anaconda.com/miniconda/ for details</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>wget<span class="w"> </span>https://repo.anaconda.com/miniconda/Miniconda3-latest-&lt;PLATFORM&gt;.sh
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>chmod<span class="w"> </span>+x<span class="w"> </span>Miniconda3-latest-&lt;PLATFORM&gt;.sh
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>./Miniconda3-latest-&lt;PLATFORM&gt;.sh
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="nb">export</span><span class="w"> </span><span class="nv">PATH</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$HOME</span><span class="s2">/miniconda3/bin:</span><span class="nv">$PATH</span><span class="s2">&quot;</span>
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="c1"># Reload shell configuration</span>
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="nb">source</span><span class="w"> </span>~/.bashrc<span class="w">  </span><span class="c1"># For Bash (Most Linux)</span>
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="nb">source</span><span class="w"> </span>~/.zshrc<span class="w">   </span><span class="c1"># For Zsh (Default in Mac)</span>
</span></code></pre></div></li>
<li>Create and activate a Conda environment
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>twilight<span class="w"> </span><span class="nv">python</span><span class="o">=</span><span class="m">3</span>.11<span class="w"> </span>-y
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>conda<span class="w"> </span>activate<span class="w"> </span>twilight
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="c1"># Set up channels</span>
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>defaults
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>bioconda
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a>conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>conda-forge
</span><span id="__span-1-7"><a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a>conda<span class="w"> </span>config<span class="w"> </span>--set<span class="w"> </span>channel_priority<span class="w"> </span>strict
</span><span id="__span-1-8"><a id="__codelineno-1-8" name="__codelineno-1-8" href="#__codelineno-1-8"></a><span class="c1"># Install TWILIGHT</span>
</span><span id="__span-1-9"><a id="__codelineno-1-9" name="__codelineno-1-9" href="#__codelineno-1-9"></a>conda<span class="w"> </span>install<span class="w"> </span>bioconda::twilight
</span></code></pre></div></li>
<li>Install TWILIGHT iterative mode
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/TurakhiaLab/TWILIGHT.git
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="nb">cd</span><span class="w"> </span>TWILIGHT
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>bash<span class="w"> </span>./install/installIterative.sh
</span></code></pre></div></li>
</ol>
<h3 id="using-installation-script-requires-sudo-access"><strong>Using installation script (requires sudo access)</strong> <a name="script"></a></h3>
<p>Users without sudo access are advised to install TWILIGHT via <a href="#conda">Conda</a> or <a href="#docker">Docker</a>.</p>
<ol>
<li>Clone the repository
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/TurakhiaLab/TWILIGHT.git
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="nb">cd</span><span class="w"> </span>TWILIGHT
</span></code></pre></div></li>
<li>Install dependencies (requires sudo access)<br />
    TWILIGHT depends on the following common system libraries, which are typically pre-installed on most development    environments:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>-<span class="w"> </span>wget
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>-<span class="w"> </span>build-essential<span class="w"> </span>
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>-<span class="w"> </span>cmake<span class="w"> </span>
</span><span id="__span-4-4"><a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>-<span class="w"> </span>libboost-all-dev<span class="w"> </span>
</span></code></pre></div>
    It also requires <code>libtbb-dev</code>, which is not always pre-installed on all systems. For users who do not have sudo     access and are missing <strong>only</strong> <code>libtbb-dev</code>, our script builds and installs TBB from source in the local user  environment, with <strong>no sudo access required</strong>.<br />
    For Ubuntu users with sudo access, if any of the required libraries are missing, you can install them with:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>sudo<span class="w"> </span>apt<span class="w"> </span>install<span class="w"> </span>-y<span class="w"> </span>wget<span class="w"> </span>build-essential<span class="w"> </span>libboost-all-dev<span class="w"> </span>cmake<span class="w"> </span>libtbb-dev
</span></code></pre></div>
    For Mac users, install dependencies using <strong>Homebrew</strong>:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>xcode-select<span class="w"> </span>--install<span class="w"> </span><span class="c1"># if not already installed</span>
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>brew<span class="w"> </span>install<span class="w"> </span>wget<span class="w"> </span>boost<span class="w"> </span>cmake<span class="w"> </span>tbb
</span></code></pre></div></li>
<li>Build TWILIGHT<br />
    Our build script automatically detects the best available compute backend <strong>(CPU, NVIDIA GPU, or AMD GPU)</strong> and builds TWILIGHT accordingly. Alternatively, users can manually specify the desired target platform.<br />
    Automatic build:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>bash<span class="w"> </span>./install/buildTWILIGHT.sh
</span></code></pre></div>
    Build for a specific platform:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>bash<span class="w"> </span>./install/buildTWILIGHT.sh<span class="w"> </span>cuda<span class="w"> </span><span class="c1"># For NVIDIA GPUs</span>
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>bash<span class="w"> </span>./install/buildTWILIGHT.sh<span class="w"> </span>hip<span class="w">  </span><span class="c1"># For AMD GPUs</span>
</span></code></pre></div></li>
<li>The TWILIGHT executable is located in the <code>bin</code> directory and can be run as follows:
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="nb">cd</span><span class="w"> </span>bin
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>./twilight<span class="w"> </span>--help
</span></code></pre></div></li>
<li><a name="install-iter"></a>(optional) Install TWILIGHT iterative mode (ensure <a href="#installconda">Conda is installed</a> first)
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="c1"># Create and activate a Conda environment </span>
</span><span id="__span-10-2"><a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>twilight<span class="w"> </span><span class="nv">python</span><span class="o">=</span><span class="m">3</span>.11<span class="w"> </span>-y
</span><span id="__span-10-3"><a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a>conda<span class="w"> </span>activate<span class="w"> </span>twilight
</span><span id="__span-10-4"><a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="c1"># Install Snakemake and tree inference tools</span>
</span><span id="__span-10-5"><a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a>bash<span class="w"> </span>./install/installIterative.sh
</span></code></pre></div></li>
</ol>
<h3 id="using-dockerfile"><strong>Using Dockerfile</strong> <a name="docker"></a></h3>
<p>The Dockerfile installed all the dependencies and tools for TWILIGHT default/iterative mode.  </p>
<ol>
<li>Clone the repository<br />
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/TurakhiaLab/TWILIGHT.git
</span><span id="__span-11-2"><a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="nb">cd</span><span class="w"> </span>TWILIGHT
</span></code></pre></div></li>
<li>Build the docker image<br />
    CPU version<br />
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="nb">cd</span><span class="w"> </span>docker/cpu
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>docker<span class="w"> </span>build<span class="w"> </span>-t<span class="w"> </span>twilight<span class="w"> </span>.
</span></code></pre></div>
    GPU version (using nvidia/cuda as base image)<br />
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="nb">cd</span><span class="w"> </span>docker/gpu
</span><span id="__span-13-2"><a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>docker<span class="w"> </span>build<span class="w"> </span>-t<span class="w"> </span>twilight<span class="w"> </span>.
</span></code></pre></div></li>
<li>Start and run docker container <br />
    CPU version
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>docker<span class="w"> </span>run<span class="w"> </span>--platform<span class="o">=</span>linux/amd64<span class="w"> </span>-it<span class="w"> </span>twilight
</span></code></pre></div>
    GPU version<br />
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>docker<span class="w"> </span>run<span class="w"> </span>--platform<span class="o">=</span>linux/amd64<span class="w"> </span>--gpus<span class="w"> </span>all<span class="w"> </span>-it<span class="w"> </span>twilight
</span></code></pre></div></li>
<li>Run TWILIGHT
    <div class="language-bash highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>./twilight<span class="w"> </span>--help
</span></code></pre></div></li>
</ol>
<h2 id="run-twilight"><a name="Run TWILIGHT"></a> <strong>Run TWILIGHT</strong></h2>
<h3 id="twilight-command-line-interface-cli"><a name="TWILIGHT CLI"></a> <strong>TWILIGHT Command Line Interface (CLI)</strong></h3>
<h4 id="main-functionalities"><strong>Main Functionalities</strong></h4>
<ol>
<li><a href="#default">Build MSA from unaligned Sequences (Default Mode)</a></li>
<li><a href="#merge-msa">Merging Multiple MSAs</a></li>
<li><a href="#add-new-sequences-1">Add New Sequences to an Existing Alignment</a></li>
</ol>
<h4 id="command-line-options"><strong>Command Line Options</strong></h4>
<div name="table1" align="center"> <b>Table 1:</b> List of command line options supported by TWILIGHT </div>

<table>
<thead>
<tr>
<th><strong>Option</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Inputs</strong></td>
<td></td>
</tr>
<tr>
<td>Build MSA from unaligned Sequences (Default Mode)</td>
<td></td>
</tr>
<tr>
<td><code>-t</code>, <code>--tree</code></td>
<td>Input guide tree file in <em>Newick format</em> (required).</td>
</tr>
<tr>
<td><code>-i</code>, <code>--sequences</code></td>
<td>Input unaligned sequences file in <em>FASTA format</em> (required).</td>
</tr>
<tr>
<td>Merging Multiple MSAs</td>
<td></td>
</tr>
<tr>
<td><code>-f</code>, <code>--files</code></td>
<td>Path to the directory containing all MSA files in <em>FASTA format</em> to be merged.</td>
</tr>
<tr>
<td>Add New Sequences to an Existing Alignment</td>
<td></td>
</tr>
<tr>
<td><code>-t</code>, <code>--tree</code></td>
<td>Input tree file in <em>Newick format</em> with placements for sequences to be aligned (optional).</td>
</tr>
<tr>
<td><code>-i</code>, <code>--sequences</code></td>
<td>Input unaligned sequences file in <em>FASTA format</em> (required).</td>
</tr>
<tr>
<td><code>-a</code>, <code>--alignment</code></td>
<td>Backbone MSAs in <em>FASTA format</em> (required).</td>
</tr>
<tr>
<td><strong>Outputs/Files</strong></td>
<td></td>
</tr>
<tr>
<td><code>-o</code>, <code>--output</code></td>
<td>Output MSA file name (required).</td>
</tr>
<tr>
<td><code>-d</code>, <code>--temp-dir</code></td>
<td>Directory for storing temporary files. Used when <code>-f</code> or <code>-m</code> is specified.</td>
</tr>
<tr>
<td><code>-k</code>, <code>--keep-temp</code></td>
<td>Keep the temporary directory (default: disabled).</td>
</tr>
<tr>
<td><code>--overwrite</code></td>
<td>Force overwriting the temporary and output files (default: disabled).</td>
</tr>
<tr>
<td><code>--write-prune</code></td>
<td>Write the pruned tree to the output directory (must be used with <code>--prune</code>, default: disabled).</td>
</tr>
<tr>
<td><strong>Hardware Usage</strong></td>
<td></td>
</tr>
<tr>
<td><code>-C</code>, <code>--cpu</code></td>
<td>Number of CPU cores (default: all available cores).</td>
</tr>
<tr>
<td>Options only for GPU version</td>
<td></td>
</tr>
<tr>
<td><code>-G</code>, <code>--gpu</code></td>
<td>Number of GPU (default: all available GPUs).</td>
</tr>
<tr>
<td><code>--gpu-index</code></td>
<td>Specify the GPU to be used, separated by commas. For example, <code>0,2,3</code> uses 3 GPUs with the index 0, 2 and 3.</td>
</tr>
<tr>
<td><code>--cpu-only</code></td>
<td>Run the program only using CPU.</td>
</tr>
<tr>
<td><strong>Alignment Parameters</strong></td>
<td></td>
</tr>
<tr>
<td><code>--type</code></td>
<td>Data type: <code>n</code> for nucleotide, <code>p</code> for protein. Automatically inferred if not specified.</td>
</tr>
<tr>
<td><code>-m</code>, <code>--max-subtree</code></td>
<td>Maximum number of leaves per subtree; enables <a href="#divide-and-conquer">divide-and-conquer method</a> (default: <code>INT_MAX</code>).</td>
</tr>
<tr>
<td><code>-r</code>, <code>--remove-gappy</code></td>
<td>Threshold for removing gappy columns (set to 1 to disable, default: 0.95).</td>
</tr>
<tr>
<td><code>--prune</code></td>
<td>Prune the input guide tree based on the presence of unaligned sequences (default: disabled).</td>
</tr>
<tr>
<td><code>-w</code>, <code>--wildcard</code></td>
<td>Treat unknown or ambiguous bases as wildcards and align them to usual letters (default: disabled).</td>
</tr>
<tr>
<td><code>--no-align-gappy</code></td>
<td>Do not align gappy columns; this will create a longer MSA (default: disabled).</td>
</tr>
<tr>
<td><code>--length-deviation</code></td>
<td>Sequences whose lengths deviate from the average by more than the specified fraction will be deferred or excluded (default: disabled).</td>
</tr>
<tr>
<td><code>--max-ambig</code></td>
<td>Sequences with an ambiguous character proportion exceeding the specified threshold will be deferred or excluded (default: 0.1).</td>
</tr>
<tr>
<td><code>--filter</code></td>
<td>Exclude sequences with high ambiguity or length deviation (default: disabled.)</td>
</tr>
<tr>
<td><code>--merge</code></td>
<td>Method to merge subtrees: <code>t</code> for Transitivity Merger and <code>p</code> for Progressive Alignment (default: <code>p</code>).</td>
</tr>
<tr>
<td><strong>Scoring Parameters</strong></td>
<td></td>
</tr>
<tr>
<td><code>--match</code></td>
<td>Match score (default: 18).</td>
</tr>
<tr>
<td><code>--mismatch</code></td>
<td>Mismatch penalty for transversions (default: -8).</td>
</tr>
<tr>
<td><code>--transition</code></td>
<td>Score for transition (default: -4).</td>
</tr>
<tr>
<td><code>--gap-open</code></td>
<td>Gap-open penalty (default: -50).</td>
</tr>
<tr>
<td><code>--gap-extend</code></td>
<td>Gap-extend penalty (default: -5).</td>
</tr>
<tr>
<td><code>-x</code>, <code>--matrix</code></td>
<td>User-specified substitution matrix path.</td>
</tr>
<tr>
<td><code>--xdrop</code></td>
<td>X-drop value (scale). The actual X-drop will be multiplied by the gap-extend penalty (default: 600).</td>
</tr>
<tr>
<td><strong>General</strong></td>
<td></td>
</tr>
<tr>
<td><code>--check</code></td>
<td>Check the final alignment. Sequences with no legal alignment will be displayed.</td>
</tr>
<tr>
<td><code>-v</code>, <code>--verbose</code></td>
<td>Print out every detail process.</td>
</tr>
<tr>
<td><code>-h</code>, <code>--help</code></td>
<td>Print help messages and exit.</td>
</tr>
<tr>
<td><code>--version</code></td>
<td>Show TWILIGHT version and exit.</td>
</tr>
</tbody>
</table>
<h4 id="usages-and-examples"><strong>Usages and Examples</strong></h4>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All files used for the examples below can be found in <code>TWILIGHT/dataset</code>.</p>
</div>
<p>Enter into the build directory (assuming <code>$TWILIGHT_HOME</code> directs to the TWILIGHT repository directory)
<div class="language-bash highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a><span class="nb">cd</span><span class="w"> </span><span class="nv">$TWILIGHT_HOME</span>/bin
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>./twilight<span class="w"> </span>-h<span class="w"> </span><span class="c1"># See Help Messages</span>
</span></code></pre></div></p>
<h5 id="build-msa-from-unaligned-sequences-default-mode"><a name="default"></a> <strong>Build MSA from unaligned Sequences (Default Mode)</strong></h5>
<p>Generate MSA by giving an unaligned sequence file and a guide tree file.  </p>
<ul>
<li>Usage syntax
<div class="language-bash highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>./twilight<span class="w"> </span>-t<span class="w"> </span>&lt;tree<span class="w"> </span>file&gt;<span class="w"> </span>-i<span class="w"> </span>&lt;sequence<span class="w"> </span>file&gt;<span class="w"> </span>-o<span class="w"> </span>&lt;output<span class="w"> </span>file&gt;
</span></code></pre></div></li>
<li>Example
<div class="language-bash highlight"><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>./twilight<span class="w"> </span>-t<span class="w"> </span>../dataset/RNASim.nwk<span class="w"> </span>-i<span class="w"> </span>../dataset/RNASim.fa<span class="w"> </span>-o<span class="w"> </span>RNASim.aln
</span></code></pre></div></li>
</ul>
<h5 id="merge-multiple-msas"><a name="merge-msa"></a> <strong>Merge Multiple MSAs</strong></h5>
<p>Assume a star-tree structure to merge multiple MSAs. Please move all MSA files to be merged into one folder.  </p>
<ul>
<li>Usage syntax
<div class="language-bash highlight"><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>./twilight<span class="w"> </span>-f<span class="w"> </span>&lt;path<span class="w"> </span>to<span class="w"> </span>the<span class="w"> </span>folder&gt;<span class="w"> </span>-o<span class="w"> </span>&lt;output<span class="w"> </span>file&gt;
</span></code></pre></div></li>
<li>Example
<div class="language-bash highlight"><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>./twilight<span class="w"> </span>-f<span class="w"> </span>../dataset/RNASim_subalignments/<span class="w"> </span>-o<span class="w"> </span>RNASim.aln
</span></code></pre></div></li>
</ul>
<h5 id="add-new-sequences-to-an-existing-alignment"><a name="add-new-sequences-1"></a>  <strong>Add New Sequences to an Existing Alignment</strong></h5>
<p>For better accuracy, it is recommended to use a tree that includes placements for the new sequences. If no tree is provided, TWILIGHT aligns new sequences to the profile of the entire backbone alignment, which may reduce accuracy. In this case, using the provided <a href="#add-new-sequences-2">Snakemake workflow</a> is advised.</p>
<ul>
<li>Usage syntax
<div class="language-bash highlight"><pre><span></span><code><span id="__span-22-1"><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>./twilight<span class="w"> </span>-a<span class="w"> </span>&lt;backbone<span class="w"> </span>alignment<span class="w"> </span>file&gt;<span class="w"> </span>-i<span class="w"> </span>&lt;new<span class="w"> </span>sequence<span class="w"> </span>file&gt;<span class="w"> </span>-t<span class="w"> </span>&lt;tree<span class="w"> </span>with<span class="w"> </span>placement<span class="w"> </span>of<span class="w"> </span>new<span class="w"> </span>sequences&gt;<span class="w"> </span>-o<span class="w"> </span>&lt;path<span class="w"> </span>to<span class="w"> </span>output<span class="w"> </span>file&gt;
</span></code></pre></div></li>
<li>Example
<div class="language-bash highlight"><pre><span></span><code><span id="__span-23-1"><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>./twilight<span class="w"> </span>-a<span class="w"> </span>../dataset/RNASim_backbone.aln<span class="w"> </span>-i<span class="w"> </span>../dataset/RNASim_sub.fa<span class="w"> </span>-t<span class="w"> </span>../dataset/RNASim.nwk<span class="w"> </span>-o<span class="w"> </span>RNASim.aln
</span></code></pre></div></li>
</ul>
<h5 id="divide-and-conquer-method"><a name="divide-and-conquer"></a> <strong>Divide-and-Conquer Method</strong></h5>
<p>Divide tree into subtrees with at most <em>m</em> leaves and align them sequentially to reduce the CPU‚Äôs main memory usage.</p>
<ul>
<li>Usage syntax
<div class="language-bash highlight"><pre><span></span><code><span id="__span-24-1"><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>./twilight<span class="w"> </span>-t<span class="w"> </span>&lt;path<span class="w"> </span>to<span class="w"> </span>tree<span class="w"> </span>file&gt;<span class="w"> </span>-i<span class="w"> </span>&lt;path<span class="w"> </span>to<span class="w"> </span>sequence<span class="w"> </span>file&gt;<span class="w"> </span>-o<span class="w"> </span>&lt;path<span class="w"> </span>to<span class="w"> </span>output<span class="w"> </span>file&gt;<span class="w"> </span>-m<span class="w"> </span>&lt;maximum<span class="w"> </span>subtree<span class="w"> </span>size&gt;
</span></code></pre></div></li>
<li>Example
<div class="language-bash highlight"><pre><span></span><code><span id="__span-25-1"><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>./twilight<span class="w"> </span>-t<span class="w"> </span>../dataset/RNASim.nwk<span class="w"> </span>-i<span class="w"> </span>../dataset/RNASim.fa<span class="w"> </span>-o<span class="w"> </span>RNASim.aln<span class="w"> </span>-m<span class="w"> </span><span class="m">200</span>
</span></code></pre></div></li>
</ul>
<h5 id="flexible-tree-support"><strong>Flexible Tree Support</strong></h5>
<p>Prunes tips that are not present in the unaligned sequence file. This is useful when working with a large tree but only aligning a subset of sequences, without needing to re-estimate the guide tree. Outputting the pruned tree is also supported.</p>
<ul>
<li>Usage syntax
<div class="language-bash highlight"><pre><span></span><code><span id="__span-26-1"><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>./twilight<span class="w"> </span>-t<span class="w"> </span>&lt;large<span class="w"> </span>tree<span class="w"> </span>file&gt;<span class="w"> </span>-i<span class="w"> </span>&lt;subset<span class="w"> </span>of<span class="w"> </span>raw<span class="w"> </span>sequences&gt;<span class="w"> </span>-o<span class="w"> </span>&lt;output<span class="w"> </span>file&gt;<span class="w"> </span>--prune<span class="w"> </span><span class="o">[</span>--write-prune<span class="o">]</span>
</span></code></pre></div></li>
<li>Example
<div class="language-bash highlight"><pre><span></span><code><span id="__span-27-1"><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>./twilight<span class="w"> </span>-t<span class="w"> </span>../dataset/RNASim.nwk<span class="w"> </span>-i<span class="w"> </span>../dataset/RNASim_sub.fa<span class="w"> </span>-o<span class="w"> </span>RNASim_sub.aln<span class="w"> </span>--prune<span class="w"> </span>--write-prune
</span></code></pre></div></li>
</ul>
<h3 id="snakemake-workflow"><a name="snakemake"></a> <strong>Snakemake Workflow</strong></h3>
<p>TWILIGHT uses the Snakemake workflow to integrate external tools for estimating guide trees and placing new sequences. These are used in iterative mode or when aligning new sequences to existing alignments. For setting up the environment and installing external tools, see <a href="#install-iter">here</a>.  </p>
<h4 id="main-functionalities_1"><strong>Main Functionalities</strong></h4>
<ol>
<li><a href="#iterative">Iterative Mode</a></li>
<li><a href="#add-new-sequences-2">Add New Sequences to an Existing Alignment (Placement Mode)</a></li>
</ol>
<h4 id="command-line-configurations"><strong>Command Line Configurations</strong></h4>
<table>
<thead>
<tr>
<th><strong>Configurations</strong></th>
<th><strong>Description and Options</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Snakemake Workflow Configurations</strong></td>
<td></td>
</tr>
<tr>
<td><code>-n</code></td>
<td>Dry run, check workflow.</td>
</tr>
<tr>
<td><code>--cores</code></td>
<td>Number of CPU cores to use (default: all available cores). Some Snakemake versions may require this to be explicitly set.</td>
</tr>
<tr>
<td><strong>User-defined Configurations (used with <code>--config</code>)</strong></td>
<td></td>
</tr>
<tr>
<td><strong>Iterative Mode</strong></td>
<td></td>
</tr>
<tr>
<td><code>SEQ</code></td>
<td>Input unaligned sequences file in <em>FASTA format</em> (required).</td>
</tr>
<tr>
<td><code>INITTREE</code></td>
<td>Tree method for initial guide tree, options: <code>parttree</code>, <code>maffttree</code> or <code>mashtree</code>. (default: <code>parttree</code>).</td>
</tr>
<tr>
<td><code>ITERTREE</code></td>
<td>Tree method for for intermediate steps, options: <code>rapidnj</code> or <code>fasttree</code>. (default: <code>rapidnj</code>).</td>
</tr>
<tr>
<td><strong>Add New Sequences to an Existing Alignment (Placement Mode)</strong></td>
<td></td>
</tr>
<tr>
<td><code>SEQ</code></td>
<td>New sequences in <em>FASTA format</em> to be placed and aligned (required).</td>
</tr>
<tr>
<td><code>ALN</code></td>
<td>Backbone alignment in <em>FASTA format</em> for placing new sequences (required.)</td>
</tr>
<tr>
<td><code>TREE</code></td>
<td>Backbone tree (optional; FastTree will be used for estimation if not provided).</td>
</tr>
<tr>
<td><strong>General</strong></td>
<td></td>
</tr>
<tr>
<td><code>TYPE</code></td>
<td>Input sequence type, options: <code>n</code> for nucleotide or <code>p</code> for protein sequences (required).</td>
</tr>
<tr>
<td><code>OUT</code></td>
<td>Output MSA file path (required).</td>
</tr>
<tr>
<td><code>DIR</code></td>
<td>Directory for storing temporary files.</td>
</tr>
<tr>
<td><code>ITER</code></td>
<td>Number of iterations (default: 3 for Iterative Mode and 2 for Placement Mode).</td>
</tr>
<tr>
<td><code>FINALTREE</code></td>
<td>Final tree estimation method (skip if unspecified), options: <code>fasttree</code>, <code>raxml</code> or <code>iqtree</code>.</td>
</tr>
<tr>
<td><code>KEEP</code></td>
<td>Keep the temporary files, options: <code>yes</code> or <code>no</code> (default: <code>no</code>).</td>
</tr>
<tr>
<td><code>OVERWRITE</code></td>
<td>Overwrite the existing files, options: <code>yes</code> or <code>no</code> (default: <code>no</code>).</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Default options in <code>workflow/config.yaml</code> will be used if the configuration is not specified. You can also modify <code>workflow/config.yaml</code> to set your options.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For users who install TWILIGHT via Conda, please replace the executable path <code>"../bin/twilight"</code> with <code>"twilight"</code> in <code>config.yaml</code>. Feel free to switch to a more powerful tree tool if available, such as replacing <code>"raxmlHPC"</code> with <code>"raxmlHPC-PTHREADS-AVX2"</code> for better performance. </p>
</div>
<h4 id="usages-and-examples_1"><strong>Usages and Examples</strong></h4>
<p>Enter <code>workflow</code> directory and type <code>snakemake</code> to view the help messages. <br />
<div class="language-bash highlight"><pre><span></span><code><span id="__span-28-1"><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a><span class="nb">cd</span><span class="w"> </span>workflow
</span><span id="__span-28-2"><a id="__codelineno-28-2" name="__codelineno-28-2" href="#__codelineno-28-2"></a>snakemake
</span><span id="__span-28-3"><a id="__codelineno-28-3" name="__codelineno-28-3" href="#__codelineno-28-3"></a><span class="c1"># or, for Snakemake versions that require specifying total number of cores:</span>
</span><span id="__span-28-4"><a id="__codelineno-28-4" name="__codelineno-28-4" href="#__codelineno-28-4"></a>snakemake<span class="w"> </span>--cores<span class="w"> </span><span class="m">1</span>
</span></code></pre></div></p>
<h5 id="iterative-mode"><a name="iterative"></a> <strong>Iterative Mode</strong></h5>
<p>TWILIGHT iterative mode estimate guide trees using external tools. </p>
<ul>
<li>Usage syntax
<div class="language-bash highlight"><pre><span></span><code><span id="__span-29-1"><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>snakemake<span class="w"> </span><span class="o">[</span>--cores<span class="w"> </span>&lt;num<span class="w"> </span>threads&gt;<span class="o">]</span><span class="w"> </span>--config<span class="w"> </span><span class="nv">TYPE</span><span class="o">=</span>VALUE<span class="w"> </span><span class="nv">SEQ</span><span class="o">=</span>VALUE<span class="w"> </span><span class="nv">OUT</span><span class="o">=</span>VALUE<span class="w"> </span><span class="o">[</span><span class="nv">OPTION</span><span class="o">=</span>VALUE<span class="w"> </span>...<span class="o">]</span>
</span></code></pre></div></li>
<li>Example - Using default configurations
<div class="language-bash highlight"><pre><span></span><code><span id="__span-30-1"><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>snakemake<span class="w"> </span>--cores<span class="w"> </span><span class="m">8</span><span class="w"> </span>--config<span class="w"> </span><span class="nv">TYPE</span><span class="o">=</span>n<span class="w"> </span><span class="nv">SEQ</span><span class="o">=</span>../dataset/RNASim.fa<span class="w"> </span><span class="nv">OUT</span><span class="o">=</span>RNASim.aln
</span></code></pre></div></li>
<li>Example - Generates the final tree based on the completed MSA.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-31-1"><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a>snakemake<span class="w"> </span>--cores<span class="w"> </span><span class="m">8</span><span class="w"> </span>--config<span class="w"> </span><span class="nv">TYPE</span><span class="o">=</span>n<span class="w"> </span><span class="nv">SEQ</span><span class="o">=</span>../dataset/RNASim.fa<span class="w"> </span><span class="nv">OUT</span><span class="o">=</span>RNASim.aln<span class="w"> </span><span class="nv">FINALTREE</span><span class="o">=</span>fasttree
</span></code></pre></div></li>
</ul>
<h5 id="add-new-sequences-to-an-existing-alignment-placement-mode"><a name="add-new-sequences-2"></a> <strong>Add New Sequences to an Existing Alignment (Placement Mode)</strong></h5>
<p>TWILIGHT aligns new sequences to the profile of the backbone alignment, infers their placement with external tools, and then refines the alignment using the inferred tree.  </p>
<ul>
<li>Usage syntax
<div class="language-bash highlight"><pre><span></span><code><span id="__span-32-1"><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>snakemake<span class="w"> </span><span class="o">[</span>--cores<span class="w"> </span>&lt;num<span class="w"> </span>threads&gt;<span class="o">]</span><span class="w"> </span>--config<span class="w"> </span><span class="nv">TYPE</span><span class="o">=</span>VALUE<span class="w"> </span><span class="nv">SEQ</span><span class="o">=</span>VALUE<span class="w"> </span><span class="nv">OUT</span><span class="o">=</span>VALUE<span class="w"> </span><span class="nv">ALN</span><span class="o">=</span>VALUE<span class="w"> </span><span class="o">[</span><span class="nv">OPTION</span><span class="o">=</span>VALUE<span class="w"> </span>...<span class="o">]</span>
</span></code></pre></div></li>
<li>Example - The backbone alignment is accompanied by a tree.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-33-1"><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a>snakemake<span class="w"> </span>--cores<span class="w"> </span><span class="m">8</span><span class="w"> </span>--config<span class="w"> </span><span class="nv">TYPE</span><span class="o">=</span>n<span class="w"> </span><span class="nv">SEQ</span><span class="o">=</span>../dataset/RNASim_sub.fa<span class="w"> </span><span class="nv">OUT</span><span class="o">=</span>RNASim.aln<span class="w"> </span><span class="nv">ALN</span><span class="o">=</span>../dataset/RNASim_backbone.aln<span class="w"> </span><span class="nv">TREE</span><span class="o">=</span>../dataset/RNASim_backbone.nwk
</span></code></pre></div></li>
<li>Example - The backbone tree is unavailable, estimate it using external tools and generate a final tree after alignment.
<div class="language-bash highlight"><pre><span></span><code><span id="__span-34-1"><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a>snakemake<span class="w"> </span>--cores<span class="w"> </span><span class="m">8</span><span class="w"> </span>--config<span class="w"> </span><span class="nv">TYPE</span><span class="o">=</span>n<span class="w"> </span><span class="nv">SEQ</span><span class="o">=</span>../dataset/RNASim_sub.fa<span class="w"> </span><span class="nv">OUT</span><span class="o">=</span>RNASim.aln<span class="w"> </span><span class="nv">ALN</span><span class="o">=</span>../dataset/RNASim_backbone.aln<span class="w"> </span><span class="nv">FINALTREE</span><span class="o">=</span>fasttree
</span></code></pre></div></li>
</ul>
<h2 id="contributions"><b>Contributions</b></h2>
<p>We welcome contributions from the community to enhance the capabilities of TWILIGHT. If you encounter any issues or have suggestions for improvement, please open an issue on <a href="https://github.com/TurakhiaLab/TWILIGHT">TWILIGHT GitHub page</a>. For general inquiries and support, reach out to our team.</p>
<h2 id="citing-twilight"><b>Citing TWILIGHT</b></h2>
<p>If you use the <b>TWILIGHT</b> in your research or publications, we kindly request that you cite the following paper:  </p>
<p>Yu-Hsiang Tseng, Sumit Walia, Yatish Turakhia, "<i>Ultrafast and ultralarge multiple sequence alignments using TWILIGHT</i>", Bioinformatics, Volume 41, Issue Supplement_1, July 2025, Pages i332‚Äìi341, doi: <a href="https://doi.org/10.1093/bioinformatics/btaf212">10.1093/bioinformatics/btaf212</a></p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/TurakhiaLab/TWILIGHT" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": ".", "features": ["content.tabs.link", "content.code.annotation", "content.code.copy", "content.action.edit", "content.action.view", "content.tooltips", "toc.follow", "toc.integrate", "navigation.tabs", "navigation.sections", "navigation.path", "navigation.tabs.sticky", "navigation.instant.prefetch", "navigation.tracking", "navigation.footer", "navigation.expand", "search.highlight", "search.suggest"], "search": "assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>